packages/remirror__extension-file/src/file-extension.tsx

```diff
@@ -0,0 +1,375 @@
+            const fileSize = anchor.getAttribute('data-filesize');
+
+            return {
+              ...extra.parse(dom),
+              url,
+              fileName,
+              fileType,
+              fileSize,
+            };
+          },
+        },
+        ...(override.parseDOM ?? []),
+      ],
+      toDOM: (node) => {
+        const { url, ...rest } = omitExtraAttributes(node.attrs, extra);
+        const attrs = {
+          ...extra.dom(node),
+          ...rest,
+          href: url,
+          'data-file': '',
+          'data-filename': node.attrs.fileName,
+          'data-filetype': node.attrs.fileType,
+          'data-filesize': node.attrs.fileSize,
+        };
+
+        return ['a', attrs];
+      },
+    };
+  }
+
+  createExternalPlugins(): ProsemirrorPlugin[] {
+    return [placeholderPlugin];
+  }
+
+  createPasteRules(): PasteRule[] {
+    return [
+      {
+        type: 'file',
+        regexp: this.options.pasteRuleRegexp,
+        fileHandler: (props) => {
+          let pos: number | undefined;
+
+          if (props.type === 'drop') {
+            pos = props.pos;
+          }
+
+          for (const file of props.files) {
+            this.startLoadFile(file, pos);
+          }
+
+          return true;
+        },
+      },
+    ];
+  }
+
+  @command()
+  uploadFile(files: File[]): CommandFunction {
+    return () => {
+      for (const file of files) {
+        this.startLoadFile(file);
+      }
+
+      return true;
+    };
+  }
+
+  @command()
+  updateFile(pos: number, attrs: FileAttributes): CommandFunction {
+    return ({ tr, dispatch }) => {
+      dispatch?.(tr.setNodeMarkup(pos, undefined, attrs));
+      return true;
+    };
+  }
+
+  @command()
+  insertFile(attributes: FileAttributes, selection?: PrimitiveSelection): CommandFunction {
+    return ({ tr, dispatch }) => {
+      const { from, to } = getTextSelection(selection ?? tr.selection, tr.doc);
+      const node = this.type.create(attributes);
+
+      dispatch?.(tr.replaceRangeWith(from, to, node));
+
+      return true;
+    };
+  }
+
+  @command()
+  deleteFile(pos: number): CommandFunction {
+    return ({ tr, state, dispatch }) => {
+      const node = state.doc.nodeAt(pos);
+
+      if (node && node.type === this.type) {
+        tr.delete(pos, pos + 1).scrollIntoView();
+        this.options.onDeleteFile({ tr, pos });
+        dispatch?.(tr);
+        return true;
+      }
+
+      return false;
+    };
+  }
+
+  @command()
+  renameFile(pos: number, fileName: string): CommandFunction {
+    return ({ tr, state, dispatch }) => {
+      const node = state.doc.nodeAt(pos);
+
+      if (node && node.type === this.type) {
+        dispatch?.(tr.setNodeMarkup(pos, undefined, { ...node.attrs, fileName }));
+        return true;
+      }
+
+      return false;
+    };
+  }
+
+  private preLoadFile(id: FileId, file: File, pos: number | undefined): Promise<FileLoader> | void {
```

Comment:
I find these method names slightly confusing, I think we're looking for something like

`preLoadFile` -> `createFilePlaceholder`

`postLoadFile` -> `onFileLoaded`

`startLoadFile` -> `uploadFile`